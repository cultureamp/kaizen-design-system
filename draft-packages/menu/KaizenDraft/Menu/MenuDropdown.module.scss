@import "~@kaizen/design-tokens/sass/border";
@import "~@kaizen/design-tokens/sass/color";
@import "~@kaizen/design-tokens/sass/shadow";
@import "~@kaizen/design-tokens/sass/spacing";

$menu-container-width: 248px;
$dropdown-index: 1000;

.menuContainer {
  z-index: $dropdown-index;
  box-sizing: border-box;
  background: $color-white;
  color: $color-purple-800;
  border-radius: $border-solid-border-radius;
  box-shadow: $shadow-large-box-shadow;
  padding: $spacing-xs 0;
  text-align: left;

  [dir="rtl"] & {
    text-align: right;
  }
}

.defaultWidth {
  width: $menu-container-width;
}
