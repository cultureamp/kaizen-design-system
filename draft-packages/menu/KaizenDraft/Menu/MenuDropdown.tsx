import * as React from "react"

const styles = require("./styles.scss")

type MenuDropdownProps = {
  hideMenuDropdown: () => void
  position?: {
    top: number
    bottom: number
    left: number
    right: number
  } | null
}

export default class MenuDropdown extends React.Component<MenuDropdownProps> {
  static displayName = "MenuDropdown"
  menu = React.createRef<HTMLDivElement>()

  componentDidMount() {
    document.addEventListener("click", this.handleDocumentClick, false)
    window.addEventListener("resize", this.handleDocumentResize, false)
    this.positionMenu()
  }

  componentWillUnmount() {
    document.removeEventListener("click", this.handleDocumentClick, false)
    window.removeEventListener("resize", this.handleDocumentResize, false)
  }

  positionMenu() {
    const menu = this.menu

    if (!this.props.position || !menu) {
      return
    }

    if (menu.current) {
      const pos = this.props.position
      const { innerHeight } = window
      const rect = menu.current.getBoundingClientRect()

      if (pos.bottom > innerHeight - rect.height) {
        menu.current.style.bottom = "24px"
        menu.current.style.top = "auto"
      } else {
        menu.current.style.top = "24px"
        menu.current.style.bottom = "auto"
      }
    }
  }

  handleDocumentClick = (e: MouseEvent) => {
    if (
      this.menu &&
      this.menu.current &&
      e.target instanceof Node &&
      !this.menu.current.contains(e.target)
    ) {
      this.props.hideMenuDropdown()
    }
  }

  handleDocumentResize = () => {
    this.props.hideMenuDropdown()
  }

  render(): JSX.Element {
    const { hideMenuDropdown, children } = this.props

    return (
      <div
        className={styles.menuContainer}
        ref={this.menu}
        onClick={() => hideMenuDropdown()}
      >
        {children}
      </div>
    )
  }
}
