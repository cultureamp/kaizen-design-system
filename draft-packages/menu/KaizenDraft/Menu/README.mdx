---
title: Menu
navTitle: Menu
summaryParagraph: A menu contains links to places or button actions. It does NOT show a selected item at all and the menu button text never changes.
tags: ["Menu list", "Context menu", "Menubar"]
needToKnow:
- If a menu dropdown contains a mix of links and buttons, separate them with a content divider with links at the top and buttons at the bottom.
demoStoryId: menu-react--label-and-icon
demoStoryHeight: 456px
---

import WhenToUseAndWhenNotToUse from "docs-components/WhenToUseAndWhenNotToUse"
import WhenToUse from "docs-components/WhenToUse"
import WhenNotToUse from "docs-components/WhenNotToUse"

## Visuals

### UI Kit

<iframe style="border: none;" width="744" height="450" src="https://www.figma.com/embed?embed_host=share&url=https%3A%2F%2Fwww.figma.com%2Ffile%2Fe3TyAOZKfLg9cA01Iy1Mgm%2FKaizen-Site-embed%3Fnode-id%3D34373%253A45952&chrome=DOCUMENTATION" allowfullscreen></iframe>

## To keep in mind

A menu provides access to additional actions and navigation links on demand while preserving a minimal interface.

*   Anatomy:
    *   The **menu button** can be any of our standard [button](/components/button) components to open the menu.
    *   The **menu list** contains a list of links, buttons, dividers, or subheading. Visually, it is bounded by a similar dropdown container to our [autocomplete](/components/autocomplete) and [select](/components/select) components. It may also be referred to as a **menu dropdown**.
*   Content:
    *   For more than 6 menu items, chunk them into sections, separated by a content [divider](/components/divider). Optionally, you may provide a subheading for each section.
    *   Destructive actions are always at the end of the menu list.
    *   For an overflow menu, use an [icon button](/components/icon-button) with a meatballs icon. Overflow menus appear on the right of page or section layouts so the menu list dropdown alignment shifts in the other direction.
    *   Use a noun for the menu button text to describe the links and actions contained within.
*   Accessibility:
    *   Ensure people using keyboards can open and close the menu by pressing the button and access links and buttons within the menu list.

## When to use and when not to use

<WhenToUseAndWhenNotToUse>
<WhenToUse>

*   Use a menu when there are link and button options available that are not used enough to show by themselves.

</WhenToUse>
<WhenNotToUse>

*   When an option from the list becomes "selected", use a [Select](/components/select) instead.
*   If there are only 2 list items, show them immediately inline instead of using a menu.
*   If you think you need 10 options, reconsider information architecture, page structure, and task flows to avoid including that many items in a menu.

</WhenNotToUse>
</WhenToUseAndWhenNotToUse>

## External Links

- [WAI-ARIA: Menu](https://www.w3.org/TR/wai-aria-1.1/#menu)
- [WAI-ARIA: Navigation menu bar example](https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-1/menubar-1.html)
- [WAI-ARIA: Editor menu bar example](https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-2/menubar-2.html)
- [WAI-ARIA: Tab panel](https://www.w3.org/TR/wai-aria-practices-1.1/#tabpanel)
- [WAI-ARIA: Toolbar](https://www.w3.org/TR/wai-aria-1.1/#toolbar)
- [MDN: Navigation role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Navigation_Role)
- [To ARIA! The Cause of, and Solution to, All Our Accessibility Problems by Jon Whiting](https://webaim.org/blog/aria-cause-solution/)

