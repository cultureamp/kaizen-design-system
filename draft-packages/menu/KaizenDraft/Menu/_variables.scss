@import "~@kaizen/design-tokens/sass/spacing";

$side-padding: $spacing-sm;
