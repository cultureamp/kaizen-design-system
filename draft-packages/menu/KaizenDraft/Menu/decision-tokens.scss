@import "~@kaizen/design-tokens/sass/theme";
@import "~@kaizen/design-tokens/sass/color";

$dt-color-border-color: $kz-color-wisteria-800;

@if $rebrand {
  $dt-color-border-color: rgba($kz-color-slate-700, 0.1);
}
