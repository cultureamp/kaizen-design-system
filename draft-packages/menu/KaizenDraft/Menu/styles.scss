@import "~@kaizen/design-tokens/sass/border";
@import "~@kaizen/design-tokens/sass/typography";
@import "~@kaizen/component-library/styles/layers";
@import "~@kaizen/deprecated-component-library-helpers/styles/type";
@import "~@kaizen/deprecated-component-library-helpers/styles/color";

$menu-container-width: 248px;

.dropdown {
  position: relative;
}

.menuContainer {
  position: absolute;
  width: $menu-container-width;
  z-index: $ca-z-index-dropdown;
  right: 0;

  [dir="rtl"] & {
    right: auto;
    left: 0;
  }
}

.reversedColor {
  color: $kz-color-white;
}
