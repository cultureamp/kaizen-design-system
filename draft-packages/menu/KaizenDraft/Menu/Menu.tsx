import { ButtonProps } from "@kaizen/draft-button"
import * as React from "react"
import MenuDropdown from "./MenuDropdown"

const styles = require("./styles.scss")

type MenuState = {
  isMenuVisible: boolean
}

export type MenuProps = {
  button: React.ReactElement<ButtonProps>
  menuVisible?: boolean
  automationId?: string
}

export default class Menu extends React.Component<MenuProps, MenuState> {
  static displayName = "Menu"
  static defaultProps = {
    iconPosition: "start",
  }

  dropdownButtonContainer = React.createRef<HTMLDivElement>()

  constructor(props: MenuProps) {
    super(props)

    this.state = {
      isMenuVisible: Boolean(props.menuVisible),
    }
  }

  toggleMenu = (e: MouseEvent) => {
    e.stopPropagation()

    const currentState = this.state.isMenuVisible
    this.setState({
      isMenuVisible: !currentState,
    })
  }

  hideMenu = () => {
    this.setState({
      isMenuVisible: false,
    })
  }

  getPosition() {
    return this.dropdownButtonContainer && this.dropdownButtonContainer.current
      ? this.dropdownButtonContainer.current.getBoundingClientRect()
      : null
  }

  renderMenuDropdown() {
    return (
      <MenuDropdown
        hideMenuDropdown={this.hideMenu}
        position={this.getPosition()}
      >
        {this.props.children}
      </MenuDropdown>
    )
  }

  render() {
    const { automationId, button } = this.props
    return (
      <div
        className={styles.dropdown}
        data-automation-id={automationId}
        ref={this.dropdownButtonContainer}
      >
        {React.cloneElement(button, {
          onClick: this.toggleMenu,
          onMouseDown: e => e.preventDefault(),
        })}
        {this.state.isMenuVisible && this.renderMenuDropdown()}
      </div>
    )
  }
}
