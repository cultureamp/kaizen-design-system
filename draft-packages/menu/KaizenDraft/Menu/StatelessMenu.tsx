import React, { useEffect, useState, SyntheticEvent, useRef } from "react"
import ReactDOM from "react-dom"
import { MenuDropdown } from "./MenuDropdown"
import styles from "./StatelessMenu.module.scss"

export type StatelessMenuProps = {
  /**
   * Whether the menu is to be used on the left or right
   * side of the viewport. If left, the left of the dropdown
   * is aligned to the left of the button (and vice versa)
   * @default "left"
   */
  align?: "left" | "right"

  /**
   * The width of the dropdown.
   * "default": a fixed width of 248px
   * "contain": contain the children's width (will be same width as children)
   * @default "default"
   */
  dropdownWidth?: "default" | "contain"

  automationId?: string
  dropdownId?: string
  /**
   * Determines when the menu should automatically hide.
   * @default: "on"
   */
  autoHide?: "on" | "outside-click-only" | "off"
  /**
   * The content to appear inside the dropdown when it is open
   */
  children: React.ReactNode
  /**
   * Render the tooltip inside a react portal, given the CSS selector.
   * This is typically used for instances where the menu is a descendant of an
   * `overflow: scroll` or `overflow: hidden` element.
   */
  portalSelector?: string
  isMenuVisible: boolean
  toggleMenuDropdown: () => void
  hideMenuDropdown: () => void
  renderButton: (args: {
    onClick: (e: any) => void
    onMouseDown: (e: any) => void
    "aria-haspopup": boolean
    "aria-expanded": boolean
  }) => React.ReactElement
  onClick?: (event: SyntheticEvent) => void
}

export const StatelessMenu = ({
  align = "left",
  dropdownWidth = "default",
  autoHide = "on",
  automationId,
  dropdownId,
  children,
  portalSelector,
  isMenuVisible,
  toggleMenuDropdown,
  hideMenuDropdown,
  renderButton,
  onClick,
}: StatelessMenuProps): JSX.Element => {
  const [referenceElement, setReferenceElement] =
    useState<HTMLSpanElement | null>(null)
  const portalSelectorElementRef = useRef<Element | null>(null)

  const menuButton = renderButton({
    onClick: (e: React.MouseEvent<Element, MouseEvent>) => {
      e.preventDefault()
      e.stopPropagation()
      toggleMenuDropdown()
    },
    onMouseDown: (e: React.MouseEvent<Element, MouseEvent>) =>
      e.preventDefault(),
    "aria-haspopup": true,
    "aria-expanded": isMenuVisible,
  })

  useEffect(() => {
    portalSelectorElementRef.current = portalSelector
      ? document.querySelector(portalSelector)
      : null
  }, [portalSelector])

  useEffect(() => {
    if (portalSelector && !portalSelectorElementRef.current) {
      // eslint-disable-next-line no-console
      console.warn(
        "The portal could not be created using the selector: " + portalSelector
      )
    }
  }, [portalSelectorElementRef, portalSelector])

  const menu = isMenuVisible ? (
    <MenuDropdown
      referenceElement={referenceElement}
      align={align}
      hideMenuDropdown={hideMenuDropdown}
      width={dropdownWidth}
      id={dropdownId}
      autoHide={autoHide}
    >
      {children}
    </MenuDropdown>
  ) : null

  return (
    // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
    <div data-automation-id={automationId} onClick={onClick}>
      <div className={styles.buttonWrapper} ref={setReferenceElement}>
        {menuButton}
      </div>
      {portalSelector && portalSelectorElementRef.current
        ? ReactDOM.createPortal(menu, portalSelectorElementRef.current)
        : menu}
    </div>
  )
}

StatelessMenu.displayName = "StatelessMenu"
