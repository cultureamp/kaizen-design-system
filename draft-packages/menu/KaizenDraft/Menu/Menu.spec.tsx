import { Button } from "@kaizen/draft-button"
import { cleanup, fireEvent, render } from "@testing-library/react"
import * as React from "react"
import Menu from "./Menu"

afterEach(cleanup)

describe("Dropdown", () => {
  it("renders default view", () => {
    const { container } = render(
      <Menu button={<Button label="Button"></Button>} />
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it("shows menu when clicking on the button", () => {
    const { container, queryByText } = render(
      <Menu button={<Button label="Button"></Button>}>
        <div>Item</div>
      </Menu>
    )

    expect(queryByText("Item")).toBeFalsy()
    const button = container.querySelector("button")
    button && fireEvent.click(button)
    expect(queryByText("Item")).toBeTruthy()
  })
})
