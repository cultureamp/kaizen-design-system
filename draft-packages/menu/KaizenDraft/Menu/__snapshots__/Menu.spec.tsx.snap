// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders default view 1`] = `
<div
  class="dropdown"
>
  <span
    class="container"
  >
    <button
      class="button"
      type="button"
    >
      <span
        class="content"
      >
        <span
          class="label"
        >
          Button
        </span>
      </span>
    </button>
  </span>
</div>
`;
